src/storage/mod.rs

```diff
@@ -568,62 +488,187 @@ impl Storage {
+                        .get(&key, &mut statistics)
+                        // map storage::txn::Error -> storage::Error
+                        .map_err(Error::from)
+                        .map(|r| {
+                            thread_ctx.collect_key_reads(CMD, 1);
+                            r
+                        });
+
+                    thread_ctx.collect_scan_count(CMD, &statistics);
+                    thread_ctx.collect_read_flow(ctx.get_region_id(), &statistics);
+
+                    result
+                })
+        }) {
+            Ok(val) => box val,
+            Err(_) => box future::err(Error::SchedTooBusy),
+        }
     }
 
+    /// Batch get from the snapshot.
     pub fn async_batch_get(
         &self,
         ctx: Context,
         keys: Vec<Key>,
         start_ts: u64,
-        callback: Callback<Vec<Result<KvPair>>>,
-    ) -> Result<()> {
-        let cmd = Command::BatchGet {
-            ctx: ctx,
-            keys: keys,
-            start_ts: start_ts,
-        };
-        let tag = cmd.tag();
-        self.schedule(cmd, StorageCb::KvPairs(callback))?;
-        KV_COMMAND_COUNTER_VEC.with_label_values(&[tag]).inc();
-        Ok(())
+    ) -> Box<Future<Item = Vec<Result<KvPair>>, Error = Error> + Send> {
+        static CMD: &'static str = "batch_get";
+        let engine = self.get_engine();
+        let priority = readpool::Priority::from(ctx.get_priority());
+
+        match self.read_pool.future_execute(priority, move |ctxd| {
+            let _timer = {
+                let ctxd = ctxd.clone();
+                let mut thread_ctx = ctxd.current_thread_context_mut();
+                thread_ctx.collect_command_count(CMD, priority, false);
+                thread_ctx.collect_command_duration(CMD)
+            };
+
+            engine
+                .future_snapshot(&ctx)
+                // map storage::engine::Error -> storage::txn::Error -> storage::Error
+                .map_err(txn::Error::from)
+                .map_err(Error::from)
+                .and_then(move |snapshot: Box<Snapshot>| {
+                    let mut thread_ctx = ctxd.current_thread_context_mut();
+                    let _t_process = thread_ctx.collect_processing_read_duration(CMD);
+
+                    let mut statistics = Statistics::default();
+                    let snap_store = SnapshotStore::new(
+                        snapshot,
+                        start_ts,
+                        ctx.get_isolation_level(),
+                        !ctx.get_not_fill_cache(),
+                    );
+                    let result = snap_store
+                        .batch_get(&keys, &mut statistics)
+                        // map storage::txn::Error -> storage::Error
+                        .map_err(Error::from)
+                        .map(|results| results
+                            .into_iter()
+                            .zip(keys)
+                            .filter(|&(ref v, ref _k)|
```

Comment:
The performance different is minimal. About the semantic, I think they have same meanings. I converted it into this way maknly for more consistent code style, because the rest of the code heavily use this "functional programming code style" and this piece code can be written in this style as well.